mhs-reference-implementation/mhs/routing/sds.py

```diff
@@ -0,0 +1,117 @@
+import asyncio
+import logging
+import ssl
+import pathlib
+from definitions import ROOT_DIR
+import ldap3
+import mhs.routing.routing_exception as re
+
+NHS_SERVICES_BASE = "ou=services, o=nhs"
+
+MHS_OBJECT_CLASS = "nhsMhs"
+AS_OBJECT_CLASS = "nhsAs"
+MHS_PARTY_KEY = 'nhsMHSPartyKey'
+
+# These need replacing as the appropriate TLS certs are received for PTL
+PRIVATE_KEY = pathlib.Path(ROOT_DIR) / 'data' / 'certs' / 'client.key'
+LOCAL_CERT_FILE = pathlib.Path(ROOT_DIR) / 'data' / 'certs' / 'client.cert'
+CA_CERTS_FILE = pathlib.Path(ROOT_DIR) / 'data' / 'certs' / 'client.cert'
+
+mhs_attributes = [
+    'nhsEPInteractionType', 'nhsIDCode', 'nhsMhsCPAId', 'nhsMHSEndPoint', 'nhsMhsFQDN',
+    'nhsMHsIN', 'nhsMHSIsAuthenticated', 'nhsMHSPartyKey', 'nhsMHsSN', 'nhsMhsSvcIA', 'nhsProductKey',
+    'uniqueIdentifier', 'nhsMHSAckRequested', 'nhsMHSActor', 'nhsMHSDuplicateElimination',
+    'nhsMHSPersistDuration', 'nhsMHSRetries', 'nhsMHSRetryInterval', 'nhsMHSSyncReplyMode'
+]
+
+
+class SDSClient:
+
+    def __init__(self, ldap_address: str = None, tls: bool = False):
+        """
+        :param ldap_address: The IP address of the directory containing data of the the MHSs
+        :param tls: Boolean defining whether or not to use TLS, this will require SSL certs to the LDAP server
+        """
+        load_tls = None
+        if ldap_address:
+            if tls:
+                load_tls = ldap3.Tls(local_private_key_file=PRIVATE_KEY,
+                                     local_certificate_file=LOCAL_CERT_FILE,
+                                     validate=ssl.CERT_REQUIRED, version=ssl.PROTOCOL_TLSv1,
+                                     ca_certs_file=CA_CERTS_FILE)
+
+            server = ldap3.Server(ldap_address, use_ssl=tls, tls=load_tls)
+            self.connection = ldap3.Connection(server, auto_bind=True, client_strategy=ldap3.REUSABLE)
+        else:
+            self.connection = None
```

Comment:
Perhaps this wasn't the correct approach but I allowed empty connections here so I could inject mocks as part of the tests without the tests erroring on instantiation. Any better ideas?

Reply:
As discussed this has been altered so that the constructor takes a connection object and null value checking is completed there. A `connectionFactory` has been implemented to produce connections and make the implementation avoid having a constructor with empty values.